File: ObdJeep/ELM327/Elm327FrameParser.swift
import Foundation

struct Elm327Frame: Equatable {
    let header: String?
    let bytes: [UInt8]
    let line: String
}

struct Elm327Response: Equatable {
    let command: Elm327Command
    let rawText: String
    let normalizedText: String
    let lines: [String]
    let frames: [Elm327Frame]
    let promptSeen: Bool
    let isOK: Bool
}

enum Elm327FrameParser {
    static func parse(rawText: String, command: Elm327Command) throws -> Elm327Response {
        let promptSeen = rawText.contains(">")
        let normalized = normalize(rawText)
        let lines = normalized
            .components(separatedBy: CharacterSet.newlines)
            .flatMap { $0.components(separatedBy: "\r") }
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty && $0 != ">" && $0 != command.command }

        if normalized.contains("UNABLE TO CONNECT") {
            throw Elm327Error.unableToConnect(raw: rawText)
        }
        if normalized.contains("BUS ERROR") || normalized.contains("BUS INIT: ERROR") {
            throw Elm327Error.busError(raw: rawText)
        }
        if normalized.contains("CAN ERROR") {
            throw Elm327Error.canError(raw: rawText)
        }
        if normalized.contains("BUFFER FULL") || normalized.contains("RX ERROR") {
            throw Elm327Error.adapterError(message: "BUFFER/RX ERROR", raw: rawText)
        }
        if normalized.contains("STOPPED") {
            throw Elm327Error.stopped(raw: rawText)
        }
        if normalized.contains("NO DATA") {
            throw Elm327Error.noData(raw: rawText)
        }
        if lines.allSatisfy({ $0 == "SEARCHING..." }) {
            throw Elm327Error.searching(raw: rawText)
        }
        if normalized.contains("?") {
            throw Elm327Error.adapterError(message: "Comando non riconosciuto", raw: rawText)
        }

        var frames: [Elm327Frame] = []
        var sawSearching = false
        for line in lines {
            if line == "SEARCHING..." {
                sawSearching = true
                continue
            }
            if line == "OK" || line.hasPrefix("ELM327") || line.hasSuffix("V") || line.contains("ISO ") || line.hasPrefix("AUTO,") {
                continue
            }
            let frame = try parseFrameLine(line, rawText: rawText)
            for index in 0..<(max(frame.bytes.count - 2, 0)) {
                if frame.bytes[index] == 0x7F {
                    throw Elm327Error.negativeResponse(
                        service: String(format: "%02X", frame.bytes[index + 1]),
                        code: String(format: "%02X", frame.bytes[index + 2]),
                        raw: rawText
                    )
                }
            }
            frames.append(frame)
        }

        if sawSearching, frames.isEmpty, !lines.contains("OK") {
            throw Elm327Error.searching(raw: rawText)
        }

        if let expected = command.expectedResponsePrefix {
            let hasExpected = normalized
                .replacingOccurrences(of: " ", with: "")
                .contains(expected.replacingOccurrences(of: " ", with: ""))
            if !hasExpected {
                throw Elm327Error.unexpectedResponse(expected: expected, raw: rawText)
            }
        }

        if frames.isEmpty && !lines.contains("OK") && !lines.contains(where: { $0.hasPrefix("ELM327") || $0.hasSuffix("V") || $0.contains("ISO ") || $0.hasPrefix("AUTO,") }) {
            throw Elm327Error.malformedFrame(raw: rawText)
        }

        return Elm327Response(
            command: command,
            rawText: rawText,
            normalizedText: normalized,
            lines: lines,
            frames: frames,
            promptSeen: promptSeen,
            isOK: lines.contains("OK")
        )
    }

    static func normalize(_ rawText: String) -> String {
        rawText
            .uppercased()
            .replacingOccurrences(of: ">", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func parseFrameLine(_ line: String, rawText: String) throws -> Elm327Frame {
        let tokens = line.components(separatedBy: CharacterSet.whitespacesAndNewlines).filter { !$0.isEmpty }
        guard !tokens.isEmpty else {
            throw Elm327Error.malformedFrame(raw: rawText)
        }

        if tokens.count == 1 {
            let bytes = try parseCompactHex(tokens[0], rawText: rawText)
            return Elm327Frame(header: nil, bytes: bytes, line: line)
        }

        var header: String?
        var bytes: [UInt8] = []
        for token in tokens {
            if token.count == 3, UInt16(token, radix: 16) != nil, header == nil {
                header = token
                continue
            }
            if token.count == 2, let byte = UInt8(token, radix: 16) {
                bytes.append(byte)
                continue
            }
            if token.count > 2, token.count.isMultiple(of: 2) {
                bytes.append(contentsOf: try parseCompactHex(token, rawText: rawText))
                continue
            }
            throw Elm327Error.malformedFrame(raw: rawText)
        }
        guard !bytes.isEmpty else {
            throw Elm327Error.malformedFrame(raw: rawText)
        }
        return Elm327Frame(header: header, bytes: bytes, line: line)
    }

    private static func parseCompactHex(_ compact: String, rawText: String) throws -> [UInt8] {
        guard compact.count >= 2, compact.count.isMultiple(of: 2) else {
            throw Elm327Error.malformedFrame(raw: rawText)
        }
        let bytes = stride(from: 0, to: compact.count, by: 2).compactMap { index -> UInt8? in
            let start = compact.index(compact.startIndex, offsetBy: index)
            let end = compact.index(start, offsetBy: 2)
            return UInt8(compact[start..<end], radix: 16)
        }
        guard bytes.count * 2 == compact.count else {
            throw Elm327Error.malformedFrame(raw: rawText)
        }
        return bytes
    }
}

File: src/core/elm327-frame-parser.ts
import { Elm327Error } from './errors'
import type { Elm327Command, Elm327Frame, Elm327Response } from './types'

const textResponses = [
  (line: string) => line === 'OK',
  (line: string) => line.startsWith('ELM327'),
  (line: string) => line.endsWith('V'),
  (line: string) => line.includes('ISO '),
  (line: string) => line.startsWith('AUTO,')
]

export class Elm327FrameParser {
  static normalize(rawText: string): string {
    return rawText.toUpperCase().replaceAll('>', '').trim()
  }

  static parse(rawText: string, command: Elm327Command): Elm327Response {
    const normalizedText = this.normalize(rawText)
    const lines = normalizedText.split(/[\r\n]+/)
      .map((line) => line.trim())
      .filter((line) => line && line !== '>' && line !== command.command)

    this.throwAdapterError(normalizedText, rawText)
    const frames: Elm327Frame[] = []
    let sawSearching = false

    for (const line of lines) {
      if (line === 'SEARCHING...') {
        sawSearching = true
        continue
      }
      if (textResponses.some((matches) => matches(line))) continue
      const frame = this.parseFrameLine(line, rawText)
      for (let index = 0; index < frame.bytes.length - 2; index += 1) {
        if (frame.bytes[index] === 0x7f) {
          throw new Elm327Error('NEGATIVE_RESPONSE', `ECU negative response service=${this.hex(frame.bytes[index + 1])} code=${this.hex(frame.bytes[index + 2])}`, rawText)
        }
      }
      frames.push(frame)
    }

    if (sawSearching && frames.length === 0 && !lines.includes('OK')) {
      throw new Elm327Error('SEARCHING', 'ELM327: SEARCHING without final response', rawText)
    }
    if (command.expectedResponsePrefix) {
      const haystack = normalizedText.replaceAll(' ', '')
      const needle = command.expectedResponsePrefix.replaceAll(' ', '').toUpperCase()
      if (!haystack.includes(needle)) throw new Elm327Error('UNEXPECTED_RESPONSE', `Expected response prefix ${needle}`, rawText)
    }
    if (frames.length === 0 && !lines.some((line) => textResponses.some((matches) => matches(line)))) {
      throw new Elm327Error('MALFORMED_FRAME', 'ELM327: malformed frame', rawText)
    }
    return { command, rawText, normalizedText, lines, frames, promptSeen: rawText.includes('>'), isOK: lines.includes('OK') }
  }

  private static throwAdapterError(normalized: string, raw: string): void {
    if (normalized.includes('UNABLE TO CONNECT')) throw new Elm327Error('UNABLE_TO_CONNECT', 'ELM327: UNABLE TO CONNECT', raw)
    if (normalized.includes('BUS ERROR') || normalized.includes('BUS INIT: ERROR')) throw new Elm327Error('BUS_ERROR', 'ELM327: BUS ERROR', raw)
    if (normalized.includes('CAN ERROR')) throw new Elm327Error('CAN_ERROR', 'ELM327: CAN ERROR', raw)
    if (normalized.includes('BUFFER FULL') || normalized.includes('RX ERROR')) throw new Elm327Error('ADAPTER_ERROR', 'ELM327: BUFFER/RX ERROR', raw)
    if (normalized.includes('STOPPED')) throw new Elm327Error('STOPPED', 'ELM327: STOPPED', raw)
    if (normalized.includes('NO DATA')) throw new Elm327Error('NO_DATA', 'ELM327: NO DATA', raw)
    if (normalized.includes('?')) throw new Elm327Error('ADAPTER_ERROR', 'ELM327: unrecognized command', raw)
  }

  private static parseFrameLine(line: string, raw: string): Elm327Frame {
    const tokens = line.split(/\s+/)
    let header: string | undefined
    const bytes: number[] = []
    for (const token of tokens) {
      if (!header && /^[0-9A-F]{3}$/.test(token)) {
        header = token
      } else if (/^[0-9A-F]{2}$/.test(token)) {
        bytes.push(Number.parseInt(token, 16))
      } else if (/^(?:[0-9A-F]{2})+$/.test(token)) {
        for (let index = 0; index < token.length; index += 2) bytes.push(Number.parseInt(token.slice(index, index + 2), 16))
      } else {
        throw new Elm327Error('MALFORMED_FRAME', 'ELM327: malformed frame', raw)
      }
    }
    if (!bytes.length) throw new Elm327Error('MALFORMED_FRAME', 'ELM327: malformed frame', raw)
    return { header, bytes, line }
  }

  private static hex(byte: number): string {
    return byte.toString(16).padStart(2, '0').toUpperCase()
  }
}
